text/0023-argo-rollouts-for-blue-green.md

```diff
@@ -0,0 +1,209 @@
+
+This generation only works under the following conditions: 
+- All services are automatically exposed via a VirtualService, i.e. Keptn cannot decide whether a service should be exposed or not
+- Istio is required
+
+**Problem statement:** The resulting drawbacks are: 
+- As the Helm chart provided by the user can only contain one deployment and one service, it is impossible to deploy multiple services together (i.e. release brackets)
+- Does not fully support edits in Git, e.g. regenerates VirtualServices when an artifact is promoted
+- Does not allow to route traffic other than HTTP
+- Does not allow to rollback to an old version, i.e. the Helm-service executes multiple Git-commits and Helm upgrades for each deployment
+- Does not allow to have different Helm charts per stage and service. This is a requirement in almost every production setting (i.e. dev has a different config than production).
+
+Overall, these drawbacks limit the blue/green deployment/releasing capabilities of the Helm-service.
+
+**Proposal:** This KEP proposes to completely remove the generation of the `generated-chart` in the helm-service and instead use [Argo Rollouts](https://argoproj.github.io/argo-rollouts/) for blue/green deployments as well as canary releases.
+
+The responsibilities of the new `helm-deploy-service` would be reduced to a Helm upgrade (i.e. `helm upgrade KEPTN_SERVICE_NAME CHART -n KEPTN_PROJECT-KEPTN_STAGE --create-namespace --install`).
+
+**Benefits:** By implementing this KEP, the following advantages can be achieved:
+- Support of Keptn service, which consists of multiple deployments, services, etc. (i.e. release brackets)
+- User can decide if a service is exposed or not
+- Argo Rollouts provide out-of-the-box canary support
+- Direct and blue/green deployments can be mixed within a project, i.e. is defined on service level
+- Removes Istio dependency
+- Easy/out-of-the-box rollback
+- Configuration changes are done via Git edits
+- GitOps approach: Work with PRs to sync stages 
+- Allows to deploy a temporary branch (i.e. developer branch)
+
+
+Furthermore, the `helm-deploy-service` should be able to work with external Helm-charts, which are managed in Helm hub, Harbor, etc. 
+Therefore, the `helm-deploy-service` should not assume that the Helm-charts are managed within Keptn's Git-repo but should support to reference external Helm-charts.
+Then it is sufficient to only manage the values.yaml file within the Keptn's Git-repo.
+
+
+## Continuous Delivery Example
+
+Next, this KEP explains how the continuous-delivery use case would look for a user.
+Therefore, let's assume a user would like to deploy a service `carts` as direct deployment in 
+`dev` and as blue/green deployment in `staging` and `production`, i.e.
+
+|       | dev    | staging    | production |
+|-------|--------|------------|------------|
+| carts | direct | blue/green | blue/green |
+
```

Comment:
Ability to combine B/G and Canary deployment.
For instance, B/G could be used for synthetic tests execution and Canary for the real traffic evaluation during the deployment bake period